compiler/passes/src/const_propagation/const_propagate_statement.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright (C) 2019-2024 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::ConstPropagator;
+
+use leo_ast::{
+    AssertStatement,
+    AssertVariant,
+    AssignStatement,
+    Block,
+    ConditionalStatement,
+    ConsoleStatement,
+    ConstDeclaration,
+    DeclarationType,
+    DefinitionStatement,
+    Expression,
+    ExpressionReconstructor,
+    ExpressionStatement,
+    IterationStatement,
+    Node,
+    ReturnStatement,
+    Statement,
+    StatementReconstructor,
+};
+
+fn empty_statement() -> Statement {
+    Statement::Block(Block { statements: Vec::new(), span: Default::default(), id: Default::default() })
+}
+
+fn is_empty_statement(stmt: &Statement) -> bool {
+    let Statement::Block(block) = stmt else {
+        return false;
+    };
+    block.statements.is_empty()
+}
+
+impl StatementReconstructor for ConstPropagator<'_> {
+    fn reconstruct_assert(&mut self, mut input: AssertStatement) -> (Statement, Self::AdditionalOutput) {
+        // It would seem natural to catch assert triggers at compile time, but this may be
+        // difficult due to control flow.
```

Comment:
In theory, one could do a final pass after all the AST transformations to catch the assertions that fail at compile time.